### PR TITLE
Split parquet file scan tasks at row group boundaries

### PR DESCRIPTION
**Pull request by samarthjain**

### CODE REVIEW

File: core/src/test/java/org/apache/iceberg/TestFixedSizeSplitScanTaskIterator.java [hunk 1]
```diff
@@ -24,9 +24,9 @@
 import org.junit.Assert;
 import org.junit.Test;
 
-import static org.apache.iceberg.BaseFileScanTask.SplitScanTaskIterator;
+import static org.apache.iceberg.BaseFileScanTask.FixedSizeSplitScanTaskIterator;
```
Comment (rdsr) [inline core/src/test/java/org/apache/iceberg/TestFixedSizeSplitScanTaskIterator.java]: It seems that `TestFixedSizeSplitScanTaskIterator `  and  `TestOffsetsBasedSplitScanTaskIterator` can be merged into one.

We will not have to duplicate `MockFileScanTask` and `verify` method.  In this case the verify method can take an extra `Iterator<FileScanTask>` argument.

Is this doable?

Comment (rdsr) [reply core/src/test/java/org/apache/iceberg/TestFixedSizeSplitScanTaskIterator.java]: The first argument in each `verify` method is only required to construct the Iterator, so we'd have to remove that too, if the Iterator is constructed outside

File: api/src/main/java/org/apache/iceberg/DataFile.java [hunk 2]
```diff
@@ -139,8 +139,10 @@ static StructType getType(StructType partitionType) {
   DataFile copy();
 
   /**
-   * @return a list of offsets for file blocks if applicable, null otherwise. When available, this
-   * information is used for planning scan tasks whose boundaries are determined by these offsets.
+   * @return List of offsets for blocks of a file, if applicable, null otherwise.
+   * When available, this information is used for planning scan tasks whose boundaries
+   * are determined by these offsets.It is important that the returned list is sorted
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/DataFile.java]: Nit: missing a space.

I think we should phrase the new content a little differently. "It is important" isn't very clear. I think it should be "offsets *must* be in sorted order."

File: api/src/main/java/org/apache/iceberg/io/FileAppender.java [hunk 3]
```diff
@@ -48,8 +48,9 @@ default void addAll(Iterable<D> values) {
   long length();
 
   /**
-   * @return a list of offsets for file blocks if applicable, null otherwise. When available, this
+   * @return a list of offsets for file blocks, if applicable, null otherwise. When available, this
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/FileAppender.java]: I missed this earlier, but why does this say "file blocks"? This should probably be "recommended split locations".

File: core/src/main/java/org/apache/iceberg/BaseFileScanTask.java [hunk 4]
```diff
@@ -70,11 +72,12 @@ public Expression residual() {
 
   @Override
   public Iterable<FileScanTask> split(long splitSize) {
-    if (file.format().isSplittable()) {
-      return () -> new SplitScanTaskIterator(splitSize, this);
-    } else {
-      return ImmutableList.of(this);
+    if (file.splitOffsets() != null) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseFileScanTask.java]: The first check should still be whether the format is splittable. If it isn't, we should not split even if there are recommended locations.

File: core/src/main/java/org/apache/iceberg/BaseFileScanTask.java [hunk 5]
```diff
@@ -86,16 +89,38 @@ public String toString() {
...
+    }
+
+    @Override
+    public boolean hasNext() {
+      return idx < splitOffsets.size();
+    }
+
+    @Override
+    public FileScanTask next() {
+      long start = splitOffsets.get(idx++);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseFileScanTask.java]: Nit: I don't think it is a good practice to use the return value of the `++` operator. It is difficult to remember whether the returned value is the original or incremented one. I think it is more clear to put this on a separate line.

File: core/src/test/java/org/apache/iceberg/TestOffsetsBasedSplitScanTaskIterator.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import com.google.common.collect.Lists;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestOffsetsBasedSplitScanTaskIterator {
+  @Test
+  public void testSplits() {
+    // case when the last row group has more than one byte
+    verify(Lists.newArrayList(4L, 10L, 15L, 18L, 30L, 45L), 48L, asList(
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestOffsetsBasedSplitScanTaskIterator.java]: Nit: `asList` can be used to make this smaller and easier to read.

File: core/src/test/java/org/apache/iceberg/TestFixedSizeSplitScanTaskIterator.java (hunk 1 above)
Comment (samarthjain) [reply core/src/test/java/org/apache/iceberg/TestFixedSizeSplitScanTaskIterator.java]: I think it makes to sense have separate tests for the two. I moved out MockFileScanTask to an outer class, instead.

Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/TestFixedSizeSplitScanTaskIterator.java]: I don't have a strong opinion on whether these should be combined. I think it's fine as long as we aren't duplicating supporting code. Sounds like that's the case now.

**Comment by rdblue** [comment]
+1 from me, and @rdsr already added a +1, so I'm going to merge this.

**Pull request merged**